zerver/views/realm_emoji.py

```diff
@@ -26,24 +27,40 @@ def upload_emoji(request: HttpRequest, user_profile: UserProfile,
                  emoji_name: Text=REQ()) -> HttpResponse:
     check_valid_emoji_name(emoji_name)
     check_emoji_admin(user_profile)
+    if RealmEmoji.objects.filter(realm=user_profile.realm,
+                                 name=emoji_name,
+                                 deactivated=False).exists():
+        return json_error(_("A realm emoji with this name already exists."))
     if len(request.FILES) != 1:
         return json_error(_("You must upload exactly one file."))
     emoji_file = list(request.FILES.values())[0]
     if (settings.MAX_EMOJI_FILE_SIZE * 1024 * 1024) < emoji_file.size:
         return json_error(_("Uploaded file is larger than the allowed limit of %s MB") % (
             settings.MAX_EMOJI_FILE_SIZE))
-    emoji_file_name = get_emoji_file_name(emoji_file.name, emoji_name)
-    upload_emoji_image(emoji_file, emoji_file_name, user_profile)
+
+    realm_emoji = check_add_realm_emoji(user_profile.realm,
+                                        emoji_name,
+                                        author=user_profile)
+    emoji_file_name = get_emoji_file_name(emoji_file.name, realm_emoji.id)
+    emoji_uploaded_successfully = False
     try:
-        check_add_realm_emoji(user_profile.realm, emoji_name, emoji_file_name, author=user_profile)
-    except ValidationError as e:
-        return json_error(e.messages[0])
+        upload_emoji_image(emoji_file, emoji_file_name, user_profile)
+        emoji_uploaded_successfully = True
+    finally:
+        if not emoji_uploaded_successfully:
+            realm_emoji.delete()
+            return json_error(_("Image file upload failed."))
+        else:
+            add_realm_emoji_filename_and_notify(realm_emoji, emoji_file_name)
```

Comment:
is there a reason for postponing addition of filename to RealmEmoji object? From my reading it seems the realmemoji object is deleted if uploading the emoji fails. But uploading success invites an extra DB operation to add file name.

Reply:
I get that you needed the realm_id for the filename but would't that be available when you do `RealmEmoji(realm=realm, name=name, author=author)` in `check_add_realm_emoji`. You can then generate the filename and store it before performing the `emoji.save()`?